bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1090 @@
+{
+	uint32_t data;
+
+	*pnt = (uint8_t *)ptr_get_be32(*pnt, &data);
+
+	return data;
+}
+
+static uint64_t pnt_decode64(uint8_t **pnt)
+{
+	uint64_t data;
+
+	*pnt = (uint8_t *)ptr_get_be64(*pnt, &data);
+
+	return data;
+}
+
+static const char *bgp_ls_print_nlri_proto(enum bgp_ls_nlri_proto proto)
+{
+	switch (proto) {
+	case BGP_LS_NLRI_PROTO_ID_IS_IS_LEVEL_1:
+		return "ISIS-L1";
+	case BGP_LS_NLRI_PROTO_ID_IS_IS_LEVEL_2:
+		return "ISIS-L2";
+	case BGP_LS_NLRI_PROTO_ID_OSPF:
+		return "OSPFv2";
+	case BGP_LS_NLRI_PROTO_ID_DIRECT:
+		return "Direct";
+	case BGP_LS_NLRI_PROTO_ID_STATIC:
+		return "Static";
+	case BGP_LS_NLRI_PROTO_ID_OSPFv3:
+		return "OSPFv3";
+	case BGP_LS_NLRI_PROTO_ID_UNKNOWN:
+		return "Unknown";
+	}
+	return "Unknown";
+}
+
+static int bgp_linkstate_nlri_node_descriptor_decode(
+	struct bgp_ls_nlri_node_descr_tlv *node_descr, uint8_t *pnt,
+	uint16_t len)
+{
+	uint8_t *lim = pnt + len;
+	uint16_t type, length;
+
+	while (pnt < lim) {
+		type = pnt_decode16(&pnt);
+		length = pnt_decode16(&pnt);
+
+		if (!bgp_ls_tlv_check_size(type, length))
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+
+		switch (type) {
+		case BGP_LS_TLV_AUTONOMOUS_SYSTEM:
+			if (CHECK_FLAG(
+				    node_descr->flags,
+				    BGP_NLRI_TLV_NODE_DESCR_AUTONOMOUS_SYSTEM))
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(node_descr->flags,
+				 BGP_NLRI_TLV_NODE_DESCR_AUTONOMOUS_SYSTEM);
+			node_descr->autonomous_system = pnt_decode32(&pnt);
+			break;
+		case BGP_LS_TLV_BGP_LS_IDENTIFIER:
+			if (CHECK_FLAG(node_descr->flags,
+				       BGP_NLRI_TLV_NODE_DESCR_BGP_LS_ID))
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(node_descr->flags,
+				 BGP_NLRI_TLV_NODE_DESCR_BGP_LS_ID);
+			node_descr->bgp_ls_id = pnt_decode32(&pnt);
+			break;
+		case BGP_LS_TLV_OSPF_AREA_ID:
+			if (CHECK_FLAG(node_descr->flags,
+				       BGP_NLRI_TLV_NODE_DESCR_AREA_ID))
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(node_descr->flags,
+				 BGP_NLRI_TLV_NODE_DESCR_AREA_ID);
+			node_descr->area_id = pnt_decode32(&pnt);
+			break;
+		case BGP_LS_TLV_IGP_ROUTER_ID:
+			if (CHECK_FLAG(node_descr->flags,
+				       BGP_NLRI_TLV_NODE_DESCR_IGP_ROUTER_ID))
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			switch (length) {
+			case BGP_LS_TLV_IGP_ROUTER_ID_ISIS_NON_PSEUDOWIRE_SIZE:
+			case BGP_LS_TLV_IGP_ROUTER_ID_ISIS_PSEUDOWIRE_SIZE:
+			case BGP_LS_TLV_IGP_ROUTER_ID_OSPF_NON_PSEUDOWIRE_SIZE:
+			case BGP_LS_TLV_IGP_ROUTER_ID_OSPF_PSEUDOWIRE_SIZE:
+				SET_FLAG(node_descr->flags,
+					 BGP_NLRI_TLV_NODE_DESCR_IGP_ROUTER_ID);
+				memcpy(&node_descr->igp_router_id, pnt, length);
+				node_descr->igp_router_id_size = length;
+				pnt += length;
+				break;
+			default:
+				zlog_err(
+					"%s: received invalid IGP Router-ID length %hu",
```

Comment:
dont return error if unknown tlv.
use conditional debugging to output the message.